### PR TITLE
feat(widgets::table): add option to always allocate the "selection" constraint

### PR DESCRIPTION
**Pull request by hasezoey**
This PR adds a option for `Table` to always allocate the "selection column", regardless of if something is selected, this way there is less "moving around" of the layout.

The default is still `false` (only allocated if something is selected), but it can both be enabled or disabled in a builder function.

Note: also renamed some internal variables to better reflect on what they do / track

### CODE REVIEW

**Comment by hasezoey** [comment]
should i do a similar thing for `List`?

File: tests/widgets_table.rs
Comment (joshka) [inline tests/widgets_table.rs:1]: I'd prefer to see unit tests (e.g. in `src/widgets/table.rs`) that just test the feature over integration tests that test the feature with a bunch of other things. (header / block / etc.). There's two reasons for this - simplicity, and reducing test failure surface.

These tests can each be about ~15 lines of code that are pure linear code (arrange, act, assert):

```rust
fn render_with_always_show_highlight_allocates_space() {
  let buffer = Buffer::empty(Rect(0, 0, 8, 3));
  let table = Table::new(vec![Row::new(vec!["a", "b", "c"])]; 2)
    .highlight_symbol(">>>")
    .highlight_show(Show::Always);
  table.render(buffer, buffer.area);
  assert_buffer_eq(
    buffer,
    Buffer::with_lines(
      "   a b c",
      "        ",
      "   a b c",
      ...
    ));
}
```

Take a look at the src/widgets/list.rs for some move examples of simplified tests.

**Review by joshka — CHANGES_REQUESTED** [review]

**Comment by joshka** [comment]
Also +1 to implementing this idea - this is otherwise pretty jarring.

File: src/widgets/table.rs
Comment (hasezoey) [reply src/widgets/table.rs]: enum added in 6e3407cfd554ce021c3eb50e869ef6ccc35d6e44, is the naming of the enum and variants OK?

File: src/widgets/table.rs
Comment (joshka) [inline src/widgets/table.rs]: Enum is named fine - the method should be more builder-like (so `highlight_spacing()`?) Don't forget to update the doc comment here too.

**Review by joshka — CHANGES_REQUESTED** [review]

File: src/widgets/table.rs
Comment (joshka) [reply src/widgets/table.rs]: In the doc comment it would be good to note why this feature exists. Sort of reiterate the commit message but reword it so that it fits well with the tone of the docs - i.e. "this works like ..." rather than "this is changed because ..."

File: src/widgets/table.rs
Comment (joshka) [inline src/widgets/table.rs:304]: You pushed just as my last comment was written :)

In case you missed it:

> In the doc comment it would be good to note why this feature exists. Sort of reiterate the commit message but reword it so that it fits well with the tone of the docs - i.e. "this works like ..." rather than "this is changed because ..."

Putting my user hat on, this comment is difficult to understand. Perhaps something like:

```suggestion
    /// When to show the highlight spacing.
    ///
    /// A highlight symbol can be displayed to the left of the row. If there is no selected row, this setting allows
    /// space to be shown or omitted. If the space is omitted then the table columns will change size when...
    pub fn highlight_spacing(mut self, value: HighlightSpacing) -> Self {
```

File: src/widgets/table.rs
Comment (hasezoey) [reply src/widgets/table.rs]: done with 21a7869ccd5e3380a0cd82826f1044ecd528e352, but had to do e2a53e1d1e3fcb543f60e8a60b26b48023a96f74 before, otherwise it would not made have proper sense

File: tests/widgets_table.rs
Comment (hasezoey) [reply tests/widgets_table.rs:1]: > that just test the feature over integration tests that test the feature with a bunch of other things. (header / block / etc.). There's two reasons for this - simplicity, and reducing test failure surface.

it is true that there should probably be unit tests instead of just integration tests, but when first looking where to add tests, i didnt find any (output testing) tests in `src/widgets/table.rs`, but in `tests/widgets_table.rs`.

also this change does need to test extra things like `header`, `block` and selection, because the column allocation also affects headers, and blocks need to be tested to ensure that the calculation is not off if there are blocks

**Comment by hasezoey** [comment]
i think i addressed all requests, is the current state OK?

PS: someone with write, could the "CI / coverage (pull_request)" workflow be re-run?

File: tests/widgets_table.rs
Comment (joshka) [reply tests/widgets_table.rs:1]: Sounds fine

**Review by joshka — APPROVED** [review]

**Comment by joshka** [comment]
I'll wait on a second review on this one.

**Comment by a-kenji** [comment]
Thanks for implementing this, this has been a pet peeve of mine on other `ratatui` apps.

I am not sure if allocating is a fitting word here, it confused me initially.
I also think that this is a breaking change, since the interface of a public struct is changing.

Otherwise LGTM.

**Comment by joshka** [comment]
> I also think that this is a breaking change, since the interface of a public struct is changing.

All the fields on `Table` are private, and the change just adds a new method, so I think this is probably in the mostly ok bucket (except for people doing "Strange Things"(TM)). What do you think?

**Comment by hasezoey** [comment]
> I am not sure if allocating is a fitting word here, it confused me initially.

any suggestions what to use instead?

**Comment by joshka** [comment]
> any suggestions what to use instead?

How about something like:

`feat(table): add option to reserve space for highlight symbol`

with a commit body like:

```text
Before this option was available, selecting an item when there was no item selected
previously made the table's layout change. This enables the table width to take up the
same amount of space whether there is a selection or not.
```

**Comment by hasezoey** [comment]
is the change in 146d7913ab36093dbe6b89647408c4c5ee34394a enough or should something else be changed?

**Comment by joshka** [comment]
Sorry - I miscommunicated here a little. The suggested text in my previous comment was for the git commit history rather than the doc comments.

I suggest the docs should be written about the current implementation, and they don't need to cover history unless there's a strong likelihood of users needing to transition. Because this is new code, I'd avoid this.

The git commit is where we talk in terms of what the change was "This is how things changed".

**Review by joshka — APPROVED** [review]

File: src/widgets/table.rs
Comment (kdheepak) [inline src/widgets/table.rs:163]:
```suggestion
/// This option allows the user to configure the "highlight symbol" column width spacing
```

File: src/widgets/table.rs
Comment (kdheepak) [inline src/widgets/table.rs:171]:
```suggestion
    /// Only add spacing for the row selection symbol if a row is selected
```

File: src/widgets/table.rs
Comment (kdheepak) [inline src/widgets/table.rs:177]:
```suggestion
    /// Never add spacing selection symbol spacing, regardless of whether something is selected or not
```

File: src/widgets/table.rs
Comment (kdheepak) [inline src/widgets/table.rs:166]:
```suggestion
    /// Always add spacing for the selection symbol column
```

File: src/widgets/table.rs
Comment (kdheepak) [inline src/widgets/table.rs:179]:
```suggestion
    /// This effectively ensures that the highlight symbol will never be drawn
```

File: src/widgets/table.rs
Comment (kdheepak) [inline src/widgets/table.rs:326]:
```suggestion
    /// See [HighlightSpacing] about which variant affects spacing in which way
```

**Comment by kdheepak** [comment]
Thanks for the PR! This is a feature that I've wanted for a while as well!

**Review by kdheepak — APPROVED** [review]
I only have minor documentation related comments. The PR LGTM otherwise.

**Comment by hasezoey** [comment]
just as a note, i have explicitly mentioned "row selection symbol" instead of just "selection symbol" for things like #376

**Comment by hasezoey** [comment]
force-pushed to rebase onto latest master (which added `Hash` on `Table`) and to fix cargo fmt for the suggestions commit

**Comment by hasezoey** [comment]
i will ask again, just to be sure:

> should i do a similar change for `List`?

**Review by orhun — APPROVED** [review]
All good, but let's use a commit message such as `feat(table): ...` before merge 🐻

**Review by joshka — APPROVED** [review]

**Comment by joshka** [comment]
> @orhun are the commits correct now? i personally use that style for my projects, to quickly note which subject (file) is meant, because there could be multiple things that have the same name in different paths (example in this project `table` as in `widgets` or as in `example`)

Keeping it broad is fine for this - it's mostly for the [Changelog](https://github.com/ratatui-org/ratatui/blob/main/CHANGELOG.md)

**Comment by joshka** [comment]
>  should i do a similar change for List?

Yes please. Thanks for this change.

**Pull request merged**
